UIE-93 Decouple data clients - Environments page pt1

msilva-broad opened the pull request:
- refactored data needs of Environments component to be aligned with Data Provider layer of the UI stack, above Data Clients (Ajax).
- Environments now requires a number of provider props, and corresponding provider implementations are implemented to abstract Ajax mechanics away.
- useWorkspaces hook decoupled with similar provider decoupling.
- addressing getUser() dependency to come later.
- BREAKING: removed all references to useReplaceableAjaxExperimental mechanic.
- implemented useModalHandler hook to streamline modal mechanics.  Partial application in Environments component modals with deleteAppModal so far.
- Also started to shift sub-components of Environments component to seperate files and improved associated types (more to come later).

### Jira Ticket: https://broadworkbench.atlassian.net/browse/UIE-93

<!-- ### Dependencies --->
<!-- Include any dependent tickets and describe the relationship. Include any relevant Jira tickets. --->

## Summary of changes:
Should have no functional change.  

All-of-us consumption would BREAK since useReplaceableAjaxExperimental is being removed.  All-of-us would need to shift to composition similar to what EnvironmentsPage wrapper component is now doing to set provider dependencies through well typed EnvironmentsProps

### What
- Replace the earlier code-sharing mechanic with the first iteration towards a UI-Stack influenced design as outlined in:  [Code Sharing Design Proposals](https://docs.google.com/document/d/1vefkF3tbIPgn_2Kbx6lscFK6nYoZoaowlkXXkszxijA/edit#heading=h.w7o97nwl0rth) .  Thought note that the Composable Dependencies design is being deferred in favor of simpler mechanisms.  Here we are using simple component props for provider dependencies to see how that plays out.  Environments component is still part of main Terra UI project.  Splitting off to a sub-package will come later.

### Why
- hoping to have this serve as the basis for the recommended strategy of dealing with data/state dependencies of large components we want to be shareable across teams.  Further refinement would include more cleanup and creation of an Analysis components package for shared Environments component (and then others), and resolving additional util/sub-dependencies that such a package split would force.

### Testing strategy
- [x] smoke test Environments page
- [x] deeper testing for app, disk, and runtime edge cases not yet covered by unit tests

<!-- ### Visual Aids -->
<!-- https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac -->

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- refactored data needs of Environments component to be aligned with Data Provider layer of the UI stack, above Data Clients (Ajax).
- Environments now requires a number of provider props, and Corresponding provider implementations are implemented to abstract Ajax mechanics away.
- useWorkspaces hook decoupled with similar provider decoupling.
- addressing getUser() dependency to come later.
- removed all references to useReplaceableAjaxExperimental mechanic.
- Also started to shift sub-components of Environments component to seperate files and improved associated types (more to come later).

jdcanas commented on src/libs/ajax/leonardo/providers/LeoRuntimeProvider.ts:
```diff
@@ -0,0 +1,28 @@
+    return Ajax(signal).Runtimes.listV2(listArgs);
+  },
+  stop: (runtime: RuntimeWrapper, signal?: AbortSignal): Promise<void> => {
+    return Ajax(signal).Runtimes.runtimeWrapper(runtime).stop();
```
perhaps it is worth removing `runtimeWrapper` from `Runtimes`, since the type `RuntimeWrapper` is more or less the same as `RuntimeBasics` and `Runtimes` has the logic the provider has within the `runtimeWrapper`.

At the very least, it would likely be worth implementing the functionality in `runtimeWrapper` here, and then leaving it for backwards compatibility until other components adopt the usage of this provider

msilva-broad replied:
I held off on a more dramatic change since quite a few consumers would be touched by changes to Ajax.Runtimes.runtimeWrapper pattern....  I am thinking this could be it's own PR to re-align all consumers to LeoRuntimeProvider and shift what's currently happening in Ajax layer to happen in this provider instead.

jdcanas replied:
Yeah, I think a TODO to align `RuntimeWrapper` and `RuntimeBasics` would be sufficient

msilva-broad replied:
added TODO

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- renamed Environments component props with new proposed naming convention.
- narrowed Environments data provider interfaces to only include needed methods to guard against the providers growing with methods used by other components but not this one.

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- rebase with dev
- switched React.FC<Props> typing on components in this PR to (props:PropsType): ReactNode typing as decided by UIWG discussion.

msilva-broad commented on src/libs/ajax/leonardo/providers/LeoAppProvider.ts:
```diff
@@ -0,0 +1,20 @@
+  pause: (resourceName: string, appName: string, signal?: AbortSignal) => Promise<void>;
+}
+
+export const leoAppProvider: LeoAppProvider = {
```
I realized I didn't include unit test coverage for these, which is needed since Environments page that uses them is mocking these more convenient interfaces in it's unit tests, so no free test coverage from there.  I will add unit tests to these providers today (Sept 7th)

msilva-broad replied:
coverage added

nawatts commented on src/analysis/Environments/DeleteAppModal.ts:
```diff
@@ -0,0 +1,66 @@
+  )(async () => {
+    // TODO: this should use types in IA-3824
+    if (cloudProvider === 'GCP') {
+      // await ajax().Apps.app(cloudResource, appName).delete(deleteDisk);
```
```suggestion
```

msilva-broad replied:
fixed

nawatts commented on src/analysis/Environments/Environments.ts:
```diff
@@ -238,29 +206,37 @@ export function PauseButton({ cloudEnvironment, currentUser, pauseComputeAndRefr
-interface DecoratedResourceAttributes {
-  workspace: WorkspaceInfo;
-  unsupportedWorkspace: boolean;
+export interface UseWorkspacesResult {
```
```suggestion
export type UseWorkspacesStateResult {
```

See below comment for name.

And nit about mixing `interface` and `type`.

msilva-broad replied:
I will argue that while they are generally interchangeable... interfaces edge out the two overall if there is a choice.

https://www.typescriptlang.org/play#example/types-vs-interfaces

nawatts replied:
I think we've had this discussion before. I prefer always using `type` because:
- It must be used for some cases: declaring a union type, etc. Thus, if we must use `type` some of the time, might as well use it all of the time and reduce the number of concepts present in the code.
- `type` behaves more intuitively. `interface` has the surprising behavior of allowing extending an interface by declaring it multiple times.

msilva-broad replied:
but how about this from that link I listed above:

// That said, we recommend you use interfaces over type
// aliases. Specifically, because you will get better error
// messages. If you hover over the following errors, you can
// see how TypeScript can provide terser and more focused
// messages when working with interfaces like Chicken.

feels like this is a big deal to help folks who are on the learning curve of Typescript, (and my own sanity even with my TS experience)

nawatts replied:
Today I learned.

The error message is a lot clearer when using an interface vs a type.

Ok, I'm sold on interfaces.

msilva-broad replied:
... and there was much rejoicing!

Great debate though.  Never be shy to question stuff like this.

nawatts commented on src/analysis/Environments/Environments.ts:
```diff
@@ -238,29 +206,37 @@ export function PauseButton({ cloudEnvironment, currentUser, pauseComputeAndRefr
-type RuntimeWithWorkspace = DecoratedResourceAttributes & Runtime;
-type DiskWithWorkspace = DecoratedResourceAttributes & PersistentDisk;
-type AppWithWorkspace = DecoratedResourceAttributes & App;
+export type UseWorkspacesProvider = (
```
```suggestion
export type UseWorkspacesState = (
```

Together with the previous comment about `UseWorkspacesResult`, since the Environments prop is named `useWorkspacesState`, can we keep the naming consistent across all the related types?

I still find "Provider" here misleading since it sounds like something that will provide "useWorkspaces", when it's actually the interface for "useWorkspaces" itself.

msilva-broad replied:
how about UseWorkspacesStateProvider which returns UseWorkspacesState?

nawatts replied:
I would interpret UseWorkspacesStateProvider which returns UseWorkspacesState as something that provides/returns a `useWorkspacesState` function.

I think both "use" and "provider" convey a simlilar meaning, so if they're both present in the name, it sounds like there's another layer involved.

nawatts commented on src/analysis/Environments/Environments.ts:
```diff
@@ -238,29 +206,37 @@ export function PauseButton({ cloudEnvironment, currentUser, pauseComputeAndRefr
+    useWorkspacesState,
     _.update('workspaces', _.flow(_.groupBy('workspace.namespace'), _.mapValues(_.keyBy('workspace.name'))))
-  )();
+  )() as UseWorkspacesResult;
```
This is not a `UseWorkspacesResult`. The `_.update` function transforms the list of workspaces into an object of:
```
{
  [billingProject]: {
    [workspaceName]: workspace,
    ...
  },
  ...
}
```

Suggest removing the `flow` and breaking this up into two statements to clarify the transformation.

It's a bit concerning that this didn't cause any type errors in the rest of the component.

msilva-broad replied:
hmmm... ok.

I am not a fan of the lack of type safety in these _ based transforms by dynamic property based predicates in _.groupBy, _.keyBy, etc.

nawatts replied:
Yeah. I think they're best used sparingly because of that. But I think the power of being able to do complex transformations in a single line can be worth it. As long as the result is stored in a typed variable to contain the spread of `any`.

```ts
const { workspaces: workspacesList } = useWorkspaces();
const workspaces: { [namespace: string]: { [name: string]: WorkspaceWrapper } } = _.flow(..., workspacesList);
```

msilva-broad replied:
fixed with new Mutate utility type.... check it out...

nawatts commented on src/analysis/Environments/Environments.ts:
```diff
@@ -350,43 +336,30 @@ export const Environments: React.FC<EnvironmentsProps> = (props) => {
-        ]
-      )
-    );
+    const wrappedPauseCompute = withErrorReporting('Error pausing compute', async () => {
```
```suggestion
    const wrappedPauseCompute = withErrorReporting('Error pausing compute', () => {
```

This function doesn't need to be async, it doesn't await anything.

msilva-broad replied:
technically it is for two reasons:
1) withErrorReporting expects an async (Promise function)
2) each of the calls to stop or pause are themselves async.

nawatts replied:
Ok, if the type for `withErrorReporting` requires it. But re: 2, a function only needs to be `async` if it's `await`ing a Promise. It does not need to be `async` to call other functions that return Promises.

nawatts commented on src/analysis/Environments/Environments.ts:
```diff
@@ -350,43 +336,30 @@ export const Environments: React.FC<EnvironmentsProps> = (props) => {
+      }
+      // default:
+      console.error('Pause is not currently implemented for azure apps');
+      return Promise.resolve();
```
```suggestion
```

This was to satisfy TS with matching return types for all cases of `cond`. With `cond` removed, it's not necessary.

msilva-broad replied:
good catch :)

nawatts commented on src/analysis/_testData/testData.ts:
```diff
@@ -824,7 +830,7 @@ export const azureRuntime: ListRuntimeItem = {
     persistentDiskId: 16902,
     region: 'eastus',
-  },
+  } satisfies AzureConfig,
```
I suspect the Runtime model types could be improved to enforce that the cloud provider for runtimeConfig matches cloudContext.

msilva-broad replied:
I did notice a few types that needed improvint/tightening.... did a bit in this PR and hope to do more soon as I get comfortable with the code area.

nawatts commented on src/components/useModalHandler.ts:
I'm compelled to grumble about mixing this into an unrelated PR. I think I like it, but there are so many other changes in Environments that I don't think I can really give this the attention it deserves and fully evaluate it.

msilva-broad replied:
grumbling duly noted... to be honest, I was spurred to create useModalHandler to untangle some of the complexity of the rather large Environments component.

nawatts commented on src/components/useModalHandler.ts:
```diff
@@ -0,0 +1,61 @@
+    maybeRender: () => {
+      // openArgs will be given value as part of open() call above,
+      // so it's safe to assume OpenArgs typing when isOpen is true
+      return isOpen && modalFactory(openArgs as OpenArgs, handler.close);
```
For cases like this where we can assume that a variable is defined, I think the non-null assertion more clearly communicates the intention and provides greater type safety.

```suggestion
      return isOpen && modalFactory(openArgs!, handler.close);
```

msilva-broad replied:
good call :)

msilva-broad replied:
fixed

nawatts reviewed (changes requested):
It looks like there's an incorrect cast in Environments.

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- added unit tests for new data providers
- addressed PR feedback.
- added Mutate type helper to improve the type cast for mutated UseWorkspacesStateResult being done on workspaces object.
- tightened some more types, including Runtime changed to narrower ListRuntimeItem used heavily in Environments component.

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- oops, forgot a couple changes.

nawatts commented on src/libs/ajax/leonardo/providers/LeoAppProvider.test.ts:
```diff
@@ -0,0 +1,97 @@
+    // Arrange
+    const ajaxMock = mockAjaxNeeds();
+    asMockedFn(ajaxMock.Apps.listWithoutProject).mockResolvedValue([]);
+    const abort = new window.AbortController();
```
Style nit: I'd drop the "window" to make things a little more concise.

msilva-broad replied:
really?  I've run into unit test harness issues if not being explicit about globals coming from window.whatever.  I've also read that it's generally a best practice.

nawatts replied:
I won't request changes over it, but I think it's unnecessary verbosity.

What issues have you run into?

Other tests work ok without `window`.
https://github.com/DataBiosphere/terra-ui/blob/c8858c5063c796380b6dc18f48f42b8411001531/src/libs/ajax/data-table-providers/WdsDataTableProvider.test.ts#L566

And Terra UI itself sometimes accesses globals without using `window`.
https://github.com/DataBiosphere/terra-ui/blob/c8858c5063c796380b6dc18f48f42b8411001531/src/libs/ajax/incremental-response/useIncrementalResponse.ts#L19

msilva-broad replied:
I've run into Typescript issues in the past, and also for edge-cases where the jest test environment lacks a specific global, having window prefix makes it easier to chase down what's wrong.  Also, it's nice context to know that AbortController is a thing coming from window, instead of some import or elsewhere.

nawatts replied:
I've seen the TypeScript issues when accessing globals that Terra UI defines, such as `ajaxOverrideUtils`. But the browser built-ins should be fine.

msilva-broad replied:
that is a good distinction to make, but I still like explicitly prefixing with window. since it aids in adding context as I noted.

nawatts reviewed (approved):
Didn't review latest changes thoroughly, but it looks like all my earlier comments were resolved.

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- added unit tests for useModalHandler

jdcanas commented on src/libs/ajax/leonardo/providers/LeoAppProvider.ts:
```diff
@@ -0,0 +1,20 @@
+  listWithoutProject: (args: Record<string, string>, signal?: AbortSignal): Promise<ListAppResponse[]> => {
+    return Ajax(signal).Apps.listWithoutProject(args);
+  },
+  delete: (resourceName: string, appName: string, deleteDisk?: boolean, signal?: AbortSignal): Promise<void> => {
```
Hmm... resourceName and appName don't make semantic sense in the context of a disk delete. I'm not entirely sure this is the correct interface to expose. The first argument is either a workspaceId or a googleProject, the former (workspaceId) corresponding to the v2 delete endpoint and the latter (googleProject) corresponding to a v1 delete endpoint. Its similar to runtimes, where we can decide based off the resource which one to call

msilva-broad replied:
hmmm... This is app delete, with disk delete being optional and driven by the existing ajax endpoint's ability to accept the deleteDisk boolean.

msilva-broad replied:
ok, from our chat, the args should be (app:AppBasics, deleteDisk? boolean), 

type AppBasics = Pick<App['CloudContext'], 'cloudResouce'>, and Pick <App, 'appName'>.  
also will probably need workspaceId to determine platform for throwing 'not-supported' error.

nawatts replied:
Nit: Instead of a boolean parameter, consider an options object or having the caller call delete app and delete disk separately.

https://alexkondov.com/should-you-pass-boolean-to-functions/
https://www.codereadability.com/what-are-javascript-options-objects/

msilva-broad replied:
I do favor named argument objects.... for all the self-documenting reasons noted in that first article... tho Typescript and modern IDE hinting, etc. do take some of the edge off with boolean args.

named args are also nice for much easier optionals where order of the args is no longer an artificial constraint on things.    I am a little hesitant to introduce an args object for just one boolean tho... I usually wait until there's at least a couple optionals, bools, or the true/false is not enough so use an enum instead scenario.  Because named args objects do tend to add weight to the caller side code.  

Because so many of these data-provider methods have signal? as last optional arg, I think we would want to run this pattern by the broader community..... do we want to have all methods like this take a single args object with all args named?  Or do we have (mainObject, optionsArgs) type thing for scenarios like this where there is a disk, or app object that is the main thing being acted on?

nawatts replied:
I find that there is almost always a better alternative to a boolean parameter.

Even when there's only one, like in this case, it's usually tangential to the main purpose of the function. And thus, it's difficult to understand what it means when reading the call. `appProvider.delete(appId)` is clear, but with `appProvider.delete(appId, true)` I have to consult the docstring (if there is one) or code to find out what that boolean is doing. An options object makes it much more clear: `appProvider.delete(appId, { deleteDisk: true })`.

Related to that, because it is also tangential to the focus of the function, I also favor putting `signal` in an options object (and have done so in [FileBrowserProvider](https://github.com/DataBiosphere/terra-ui/blob/dev/src/libs/ajax/file-browser-providers/FileBrowserProvider.ts)).

msilva-broad replied:
I very much agree the readability benefits greatly... I saw some nice efficiencies in how you implemented the signatures too, which I will borrow some of... gonna give this a whirl...

msilva-broad replied:
@nawatts take a look at the new provider methods signatures :)

nawatts replied:
Nice!

jdcanas commented on src/libs/ajax/leonardo/providers/LeoAppProvider.ts:
```diff
@@ -0,0 +1,20 @@
+export interface LeoAppProvider {
+  listWithoutProject: (args: Record<string, string>, signal?: AbortSignal) => Promise<ListAppResponse[]>;
+  delete: (resourceName: string, appName: string, deleteDisk?: boolean, signal?: AbortSignal) => Promise<void>;
+  pause: (resourceName: string, appName: string, signal?: AbortSignal) => Promise<void>;
```
I don't think either delete or pause have the right interface (see comment below on delete)

msilva-broad replied:
fixed these up based on our convorsation.

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- uptick minor version on core-utils package

msilva-broad commented on src/libs/ajax/leonardo/providers/LeoDiskProvider.ts:
```diff
@@ -0,0 +1,24 @@
+    const { cloudContext, name, id } = disk;
+
+    if (isGcpContext(cloudContext)) {
+      const { googleProject } = workspace as GoogleWorkspaceInfo;
```
googleProject should come from disk object not workspace!!!!

msilva-broad replied:
oh no wait.... @jdcanas see my slack msg... I remember now that this was intentional on my part because I think I found a bug in the previous code... googleProject doesn't exist on PersistentDisk type...

msilva-broad replied:
ok, after digging in with @jdcanas we did indeed find a bug with the original code.  My next commit will fix this with Justin's guidance to get "goggleProject" from disk.loudContext.cloudResource, which will remove the need for workspacea arg :)

msilva-broad replied:
bug fixed.  @jdcanas please include in your testing.

msilva-broad commented on src/analysis/Environments/Environments.ts:
```diff
@@ -128,57 +145,6 @@ const DeleteDiskModal = ({ disk: { cloudContext, googleProject, name, id }, isGa
-      await ajax().Apps.app(cloudResource, appName).delete(deleteDisk);
-      onSuccess();
-    } else {
-      throw new Error('Deleting apps is currently only supported on GCP');
```
this not-supported error needs to get into AppProvider.delete

msilva-broad committed:
UIE-93 Decouple data clients - Environments page pt1
- more PR feedback and existing bug fix
- DeleteDiskModal, and related leoDiskProvider now gets resourceName (aka googleProject) from disk.cloudContext instead of a non-existent 'googleProject' property off of disk object.
- new leo providers now have methods that use option-object pattern for better code readability on booleans like deleteDisk option, included signal (cancel) option in this pattern too.

nawatts commented on src/pages/EnvironmentsPage.ts:
```diff
@@ -1,8 +1,14 @@
+import { ReactNode } from 'react';
 import { h } from 'react-hyperscript-helpers';
-import { EnvironmentNavActions, Environments } from 'src/analysis/Environments/Environments';
+import { EnvironmentNavActions, Environments, UseWorkspacesState } from 'src/analysis/Environments/Environments';
```
Shouldn't `UseWorkspaceState` be a shared interface similar to various providers?

msilva-broad replied:
yes, but didn't want to take on converting the workspace utils file it currently comes from to TS since this PR was already on the big side.  But I can do that as a small followup....

LizBaldo commented:
![Screenshot 2023-09-21 at 3 04 21 PM](https://github.com/DataBiosphere/terra-ui/assets/13465644/92dd76ff-00fb-4cd0-82e8-8511c29fb420)

FYI @msilva-broad I was trying to test this on my BEE and I am having trouble seeing the cloud environment page, could you double check?

msilva-broad commented:
> ![Screenshot 2023-09-21 at 3 04 21 PM](https://user-images.githubusercontent.com/13465644/269727975-92dd76ff-00fb-4cd0-82e8-8511c29fb420.png)
> 
> FYI @msilva-broad I was trying to test this on my BEE and I am having trouble seeing the cloud environment page, could you double check?

taking a look

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- trying fix for missing workspace on cromwell apps.

nawatts commented on packages/core-utils/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@terra-ui-packages/core-utils",
-  "version": "0.0.6",
+  "version": "0.1.6",
```
If we're going to bump this version, let's go to either `0.0.7` or `0.1.0`.

msilva-broad replied:
major = breaking change
minor = additive (non-breaking) change
patch = bugfix...

Since PR is adding stuff, minor version bump...

nawatts replied:
Yes, but there's no need to go straight to 0.1.6. Bumping one version part can reset the following parts to zero.

nawatts replied:
According to the SemVer spec, a minor version increase _must_ reset the patch version to 0.
https://semver.org/#spec-item-7

msilva-broad replied:
ok fixed, and thx for checking the homework :)

lucymcnatt commented:
Found a similar bug to last time when trying to delete a runtime from the cloud env page `Cannot read properties of undefined (reading 'workspaceId')`. This is the line the console points to `workspaceId: pe.workspace.workspaceId` and here's some context around that:

>                    diskName: _e
>                 }, g), ()=>"Cannot delete this disk because it is attached. You must delete the cloud environment first."]);
>                 return t.h(ae, {
>                     disabled: !!Pe,
>                     tooltip: Pe || "Delete persistent disk",
>                     onClick: ()=>P(ve)
>                 }, [ln("trash"), "Delete"])
>             }
>         }]
>     })]), C && t.h(av, {
>         runtime: u.find({
>             id: C
>         }, f),
>         onDismiss: ()=>w(void 0)
>     }), x && pe && t.h(jN, {
>         runtime: pe,
>         workspaceId: pe.workspace.workspaceId,
>         deleteProvider: s,
>         onDismiss: ()=>E(void 0),
>         onSuccess: ()=>{
>             E(void 0),
>             M()
>         }

lucymcnatt commented:
Another one when I try to pause a cromwell app `Cannot read properties of undefined (reading 'cloudPlatform')`
The console points me to  ` , Ij = e=>e.cloudPlatform === "Gcp"` and some context is:

> }
>   , Tq = {
>     AZURE: "Microsoft Azure",
>     GCP: "Google Cloud Platform"
> }
>   , Eq = e=>e in ai
>   , Ij = e=>e.cloudPlatform === "Gcp"
>   , D3 = ["NO ACCESS", "READER", "WRITER", "OWNER", "PROJECT_OWNER"]
>   , Hp = (e,t)=>D3.indexOf(t) >= D3.indexOf(e)
>   , Iv = e=>Hp("WRITER", e)
>   , Pq = e=>Hp("READER", e)
>   , Rq = e=>Hp("OWNER", e)
>

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- fixed more missing workspace bugs, this time with delete runtime and pause app.
- corrected type of DecoratedResourceAttributes which the code decorates runtimes, disks, and apps with.  It now has workspace as optional since it clearly can be undefined.
- the exposed the same issues these bugs encountered, which are  now addressed.
- DeleteRuntimeModal doesn't even need workspaceId anymore, so just removed that from the caller.

lucymcnatt reviewed (approved):
Manual testing is all successful now!

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- fixed package version on core-utils

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- fixed package version on core-utils (oops, fixed to 0.1.0)
